Suppress output tab setting

gonardfreeman opened the pull request:
<!--- PR title should follow the pattern: <type>(optional scope): <description>.
please refer to the types and format here: https://www.conventionalcommits.org/en/v1.0.0/#summary
If this is a feat/fix, add the technical writer as a reviewer to the PR. --->

### What does this PR do?

Added a setting that will suppress output on deploy/retreive

### What issues does this PR fix or reference?

#4970

### Functionality Before

no such setting. Messages is constantly opening bottom panel and irritating users

### Functionality After

Peace and calm after you enable it.

gonardfreeman commented:
Any feedback?

CristiCanizales commented on packages/salesforcedx-utils-vscode/src/settings/sfdxSettingsService.ts:
```diff
@@ -1,10 +1,20 @@
       .get<boolean>(SETTING_CLEAR_OUTPUT_TAB, false);
   }
+
+  public static getEnableSuppressOutputAfterEachCommand(): boolean {
```
```suggestion
  public static getEnableSuppressOutputAfterSuccessfulOperation(): boolean {
```

CristiCanizales commented on packages/salesforcedx-utils-vscode/test/unit/settings/sfdxSettingsService.test.ts:
```diff
@@ -11,14 +11,17 @@ describe('SfdxSettingsService', () => {
         SfdxSettingsService.getEnableClearOutputBeforeEachCommand()
       ).equals(true);
+      expect(
+        SfdxSettingsService.getEnableSuppressOutputAfterEachCommand()
```
```suggestion
        SfdxSettingsService.getEnableSuppressOutputAfterSuccessfulOperation()
```

CristiCanizales commented on packages/salesforcedx-utils-vscode/test/unit/settings/sfdxSettingsService.test.ts:
```diff
@@ -27,6 +30,9 @@ describe('SfdxSettingsService', () => {
         SfdxSettingsService.getEnableClearOutputBeforeEachCommand()
       ).equals(false);
+      expect(
+        SfdxSettingsService.getEnableSuppressOutputAfterEachCommand()
```
```suggestion
        SfdxSettingsService.getEnableSuppressOutputAfterSuccessfulOperation()
```

CristiCanizales commented on packages/salesforcedx-vscode-core/src/commands/baseDeployRetrieve.ts:
```diff
@@ -148,8 +149,10 @@ export abstract class DeployExecutor<T> extends DeployRetrieveExecutor<T> {
         const relativePackageDirs = await SfdxPackageDirectories.getPackageDirectoryPaths();
-        const output = this.createOutput(result, relativePackageDirs);
-        channelService.appendLine(output);
+        if (!SfdxSettingsService.getEnableSuppressOutputAfterEachCommand()) {
```
```suggestion
        if (!SfdxSettingsService.getEnableSuppressOutputAfterSuccessfulOperation()) {
```

CristiCanizales commented on packages/salesforcedx-vscode-core/src/commands/baseDeployRetrieve.ts:
```diff
@@ -278,8 +281,10 @@ export abstract class RetrieveExecutor<T> extends DeployRetrieveExecutor<T> {
       const relativePackageDirs = await SfdxPackageDirectories.getPackageDirectoryPaths();
-      const output = this.createOutput(result, relativePackageDirs);
-      channelService.appendLine(output);
+      if (!SfdxSettingsService.getEnableSuppressOutputAfterEachCommand()) {
```
```suggestion
      if (!SfdxSettingsService.getEnableSuppressOutputAfterSuccessfulOperation()) {
```

CristiCanizales commented on packages/salesforcedx-vscode-core/src/settings/sfdxCoreSettings.ts:
```diff
@@ -94,6 +95,10 @@ export class SfdxCoreSettings {
     return this.getConfigValue(SETTING_CLEAR_OUTPUT_TAB, false);
   }
 
+  public getEnableSuppressOutputAfterEachCommand(): boolean {
```
```suggestion
  public getEnableSuppressOutputAfterSuccessfulOperation(): boolean {
```

CristiCanizales commented on packages/salesforcedx-vscode-core/test/vscode-integration/commands/util/sfdxCommandlet.test.ts:
```diff
@@ -145,4 +145,54 @@ describe('SfdxCommandlet', () => {
+
+  it('Should suppress message if user preference is set to true', async () => {
+    sandbox
+      .stub(sfdxCoreSettings, 'getEnableSuppressOutputAfterEachCommand')
```
```suggestion
      .stub(sfdxCoreSettings, 'getEnableSuppressOutputAfterSuccessfulOperation')
```

CristiCanizales commented on packages/salesforcedx-vscode-core/test/vscode-integration/commands/util/sfdxCommandlet.test.ts:
```diff
@@ -145,4 +145,54 @@ describe('SfdxCommandlet', () => {
+
+  it('Should suppress message if user preference is set to false', async () => {
+    sandbox
+      .stub(sfdxCoreSettings, 'getEnableSuppressOutputAfterEachCommand')
```
```suggestion
      .stub(sfdxCoreSettings, 'getEnableSuppressOutputAfterSuccessfulOperation')
```

CristiCanizales reviewed (changes requested):
Hi @gonardfreeman! Just a few suggestions mostly regarding naming, and you would also need to make the jest test for your change in baseDeployRetrieve.ts. Thanks for this!

CristiCanizales commented on packages/salesforcedx-vscode-core/src/commands/baseDeployRetrieve.ts:
```diff
@@ -14,6 +14,7 @@ import {
   SourceTrackingType,
   Table
 } from '@salesforce/salesforcedx-utils-vscode';
+import { SfdxSettingsService } from '@salesforce/salesforcedx-utils-vscode/src/settings';
```
```suggestion
```

gonardfreeman commented on packages/salesforcedx-vscode-core/test/vscode-integration/commands/util/sfdxCommandlet.test.ts:
```diff
@@ -145,4 +145,54 @@ describe('SfdxCommandlet', () => {
+    );
+    await commandlet.run();
+    // tslint:disable-next-line:no-unused-expression
+    expect(clearStub.called).to.be.false;
```
Not sure about this one because, if I understood logic correct, this line checks if stub was called. And with setting enabled it should check if output tab is closed

gonardfreeman commented:
@CristiCanizales I've accepted all changes except one. I'm doing it for the first time, should I create another pull request or accepting from UI is enough?

gonardfreeman commented:
@CristiCanizales sorry for spam, I've added some commits to this PR because there were bug (found during unit test creation) and added unit tests for new functionality.
Sorry once again 😅

gonardfreeman commented:
To anyone who will read it 😃 

I've fixed few tests. Also I've resolved all issue that was mentioned.

CristiCanizales commented:
Hello @gonardfreeman! I haven't had the chance to take a look at your changes, that's why I haven't reached back, but be sure I'm aware of them and as soon as I do I'll let you know. Thanks!

gonardfreeman commented:
No worries @CristiCanizales 
As  manager from my first job used to say: “work must lay down a bit” the meaning is that when something is done you should give it a time before share with others. In our case I had a chance to improved tests😂

daphne-sfdc commented:
Hi @gonardfreeman,

I'm doing your code review and would like a clarification from you about the expected behavior.

I cloned your fork and used it to Launch Extensions.  In the window that popped up, I went into Settings and checked the box next to 'Salesforcedx-vscode-core: Suppress Output Tab'.  
<img width="777" alt="Screenshot 2023-08-17 at 12 44 52 PM" src="https://github.com/forcedotcom/salesforcedx-vscode/assets/139700604/78c2f624-621e-4d9e-9d6d-9839defee130">

I then opened an existing project, created a scratch org, clicked the X next to the output tab to close it, and deployed my files successfully to that scratch org.  After deploying, the output tab and this notification popped up:
<img width="1268" alt="Screenshot 2023-08-17 at 12 48 26 PM" src="https://github.com/forcedotcom/salesforcedx-vscode/assets/139700604/08812859-a26a-46a2-af75-3eafbff2cc93">

If I uncheck the box next to 'Salesforcedx-vscode-core: Suppress Output Tab' in Settings, and then deploy the same file to my scratch org, I see the output tab pop up with this:
<img width="1269" alt="Screenshot 2023-08-17 at 1 03 34 PM" src="https://github.com/forcedotcom/salesforcedx-vscode/assets/139700604/9219c20a-4d30-4cee-9792-053663600edb">

I want to check with you because you said your expected behavior is for the Output tab to not pop up at all, which does not match what I am seeing.  What I see is that the list of files that are successfully deployed is no longer listed, but the tab itself still pops up with the start and end messages.

Please help to clarify what your new setting is supposed to do so that I can better understand and continue my code review.  Thanks!

gonardfreeman commented:
Hi @daphne-sfdc !
Thanks for noticing it.
I thought that whole output is created in one chunk, but now I'll see that there is other part of that thing (start and end deploy). I'll fix it and deploy changes.
Thank you for your time looking at it.

gonardfreeman commented:
Hi @daphne-sfdc !
Sorry for such delay, got a new job and haven't had a chance to fix that. I've pushed new changes so whenever you'll get a chance please review.
Have a nice day :)

CristiCanizales commented:
Hi @gonardfreeman, I'm testing your setting and in my case the output tab is popping up both with the setting enabled and disabled... 
<img width="896" alt="image" src="https://github.com/forcedotcom/salesforcedx-vscode/assets/113132642/868cd193-6a5d-44d2-a8a6-dbcce398fefd">

gonardfreeman commented:
Hi @CristiCanizales !
[here is video](https://youtu.be/toz_irHkxVM) of how it is working and how it is intended to work. Also output will be still in the panel, but panel itself won't open.
I guess steps to reproduce will look like this:

1. enable setting
2. close bottom panel
3. deploy source to org in any way you like from VS Code (right click in file -> deploy, CTRL/CMD + Shift + P -> deploy source, etc)

gonardfreeman commented:
@CristiCanizales  I swapped tests in place, because they are failing for windows without any reason. I assume Sandbox is cached between tests or something. If it won't help, could you please point me on how to run those tests locally so I can debug them

gonardfreeman commented:
@daphne-sfdc @CristiCanizales I've fixed test issue, but also found issue in other test that is not mine.

this test 
https://github.com/forcedotcom/salesforcedx-vscode/blob/a14bbfc8e74a56dc2b98713b0b1c6a864f606e7d/packages/salesforcedx-vscode-core/test/vscode-integration/commands/util/sfdxCommandlet.test.ts#L99
and
https://github.com/forcedotcom/salesforcedx-vscode/blob/a14bbfc8e74a56dc2b98713b0b1c6a864f606e7d/packages/salesforcedx-vscode-core/test/vscode-integration/commands/util/sfdxCommandlet.test.ts#L124
this one is checking both for command not to be called, but one of them should check if it was called and other if not. I can fix it but I'll do it in other branch

daphne-sfdc commented:
Hi @gonardfreeman,
Our team is actually in the process of phasing out the integration tests, so it's ok if your test doesn't work.  I was going to push a commit to revert it but it seems like you got there to make the fix before me. 🙂 If the integration tests continue to fail, please go ahead and revert it.

We do need more unit test coverage on this PR, so since you're in the effort of creating tests, please redirect your efforts to increasing test coverage there.  Thanks.

gonardfreeman commented:
Hi @daphne-sfdc !
I've reverted changes for integration tests, they were still failing. It will be much easier if I find out how to debug them on local machine, but I had no luck with it :( 
I'm going to work with unit tests coverage on this PR. Later today I hope to push something :)

daphne-sfdc commented:
Hi @gonardfreeman,
Our team has decided that we are not planning to merge this PR due to some inconsistencies in the resulting behavior that are due to the way the extensions are originally built.  But your code changes are a step in the right direction, so we have saved them in the **pr/4971** branch to hopefully reuse once we have the bandwidth to address those inconsistencies.

Here is the background behind our decision: During the PR review process, I did some in-depth testing on your code changes, as our customary practice.  During the testing, I discovered that your setting actually suppresses the output tab for about 60% of SFDX commands (including commands like Run Apex Tests and Diff Source) instead of being confined to deploy/retrieve.  This is because your code changes were done in the LibraryCommandletExecutor, which is shared among all those commands.  (That's a good thing - we do want Suppress Output Tab to apply to all those other commands!)  However, because the remaining 40% of commands won't receive the effects of this setting, our team has decided that we would table this PR until we refactor our code to have all the commands use the LibraryCommandletExecutor and remove the inconsistencies.

Thank you for your contribution, and sorry about this update.

gonardfreeman commented:
Hi @daphne-sfdc !
No worries, thank you for the updates!
